File: backend/app/db/seed.py
"""Development seed data: demo tenants and one user per role."""

import logging

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import hash_password
from app.models import Tenant, User, UserRole
from services.account_provisioning import ensure_tenant_account_catalog, provision_trader_demo_account

logger = logging.getLogger(__name__)

DEMO_PASSWORD = "demo-password-123"  # noqa: S105 - development seed only

SEED_TENANTS = [
    {
        "slug": "app",
        "client_key": "proppredict",
        "name": "PropPredict",
        "tagline": "Trade predictions. Get funded.",
        "branding": {
            "accent": "#22c55e",
            "accent_hover": "#16a34a",
            "accent_soft": "rgba(34, 197, 94, 0.12)",
            "accent_foreground": "#04170b",
            "logo_glyph": "P",
        },
        "features": {"leaderboard": True, "journal": True, "payouts": True},
        "program": {
            "currency": "USD",
            "account_sizes": [10_000, 25_000, 50_000, 100_000],
            "profit_target_pct": 10,
            "max_daily_loss_pct": 5,
            "max_drawdown_pct": 10,
            "drawdown_mode": "static",
            "profit_split_pct": 80,
            "max_stake_per_order": 2_500,
            "max_exposure_per_market": 5_000,
            "challenge_duration_days": 60,
            "min_trading_days": 10,
        },
    },
    {
        "slug": "apex",
        "client_key": "apex",
        "name": "Apex Forecast",
        "tagline": "Elite forecasting capital.",
        "branding": {
            "accent": "#38bdf8",
            "accent_hover": "#0ea5e9",
            "accent_soft": "rgba(56, 189, 248, 0.12)",
            "accent_foreground": "#04121b",
            "logo_glyph": "A",
        },
        "features": {"leaderboard": True, "journal": True, "payouts": False},
        "program": {
            "currency": "USD",
            "account_sizes": [25_000, 50_000, 200_000],
            "profit_target_pct": 8,
            "max_daily_loss_pct": 4,
            "max_drawdown_pct": 8,
            "drawdown_mode": "trailing",
            "profit_split_pct": 90,
            "max_stake_per_order": 5_000,
            "max_exposure_per_market": 10_000,
            "challenge_duration_days": 45,
            "min_trading_days": 7,
        },
    },
    {
        "slug": "nova",
        "client_key": "nova",
        "name": "Nova Markets",
        "tagline": "Predict boldly. Trade funded.",
        "branding": {
            "accent": "#a78bfa",
            "accent_hover": "#8b5cf6",
            "accent_soft": "rgba(167, 139, 250, 0.12)",
            "accent_foreground": "#120a24",
            "logo_glyph": "N",
        },
        "features": {"leaderboard": True, "journal": False, "payouts": True},
        "program": {
            "currency": "USD",
            "account_sizes": [10_000, 50_000, 100_000],
            "profit_target_pct": 12,
            "max_daily_loss_pct": 5,
            "max_drawdown_pct": 12,
            "drawdown_mode": "static",
            "profit_split_pct": 75,
            "max_stake_per_order": 2_000,
            "max_exposure_per_market": 4_000,
            "challenge_duration_days": 90,
            "min_trading_days": 12,
        },
    },
]


async def seed_database(db: AsyncSession) -> None:
    result = await db.execute(select(Tenant.id))
    if result.first() is not None:
        return

    logger.info("Seeding demo tenants and users")
    for spec in SEED_TENANTS:
        tenant = Tenant(**spec)
        db.add(tenant)
        await db.flush()

        include_kalshi = tenant.slug == "apex"
        await ensure_tenant_account_catalog(db, tenant, include_kalshi=include_kalshi)

        db.add_all(
            [
                User(
                    tenant_id=tenant.id,
                    email=f"trader@{tenant.slug}.proppredict.com",
                    display_name="Demo Trader",
                    hashed_password=hash_password(DEMO_PASSWORD),
                    role=UserRole.TRADER,
                ),
                User(
                    tenant_id=tenant.id,
                    email=f"admin@{tenant.slug}.proppredict.com",
                    display_name="Firm Admin",
                    hashed_password=hash_password(DEMO_PASSWORD),
                    role=UserRole.PROP_FIRM_ADMIN,
                ),
            ]
        )
        await db.flush()

        trader_result = await db.execute(
            select(User).where(
                User.tenant_id == tenant.id,
                User.email == f"trader@{tenant.slug}.proppredict.com",
            )
        )
        trader = trader_result.scalar_one()

        product = None
        if include_kalshi:
            from app.models import PropFirmAccount

            kalshi_product = await db.execute(
                select(PropFirmAccount).where(
                    PropFirmAccount.tenant_id == tenant.id,
                    PropFirmAccount.slug == "kalshi",
                )
            )
            product = kalshi_product.scalar_one_or_none()

        await provision_trader_demo_account(
            db,
            user=trader,
            tenant=tenant,
            prop_firm_account=product,
        )

    first = await db.execute(select(Tenant).where(Tenant.slug == "app"))
    app_tenant = first.scalar_one()
    db.add(
        User(
            tenant_id=app_tenant.id,
            email="root@proppredict.com",
            display_name="Super Admin",
            hashed_password=hash_password(DEMO_PASSWORD),
            role=UserRole.SUPER_ADMIN,
        )
    )
    await db.commit()

File: lib/seed.ts
import { prisma } from "@/lib/db";
import { hashPassword } from "@/lib/auth";
import { createPropFirmApiKey } from "@/lib/provisioning/api-keys";
import { ensureFirmSettingsSeeded } from "@/lib/provisioning/firm-settings";

const DEMO_PASSWORD = "demo-password-123";

const SEED_TENANTS = [
  {
    slug: "app",
    clientKey: "proppredict",
    name: "PropPredict",
    tagline: "Trade predictions. Get funded.",
    branding: {
      accent: "#22c55e",
      accentHover: "#16a34a",
      accentSoft: "rgba(34, 197, 94, 0.12)",
      accentForeground: "#04170b",
      logoGlyph: "P",
    },
    features: { leaderboard: true, journal: true, payouts: true },
    program: {
      currency: "USD",
      accountSizes: [10_000, 25_000, 50_000, 100_000],
      profitTargetPct: 10,
      maxDailyLossPct: 5,
      maxDrawdownPct: 10,
      drawdownMode: "static",
      profitSplitPct: 80,
      maxStakePerOrder: 2_500,
      maxExposurePerMarket: 5_000,
      challengeDurationDays: 60,
      minTradingDays: 10,
    },
  },
  {
    slug: "apex",
    clientKey: "apex",
    name: "Apex Forecast",
    tagline: "Elite forecasting capital.",
    branding: {
      accent: "#38bdf8",
      accentHover: "#0ea5e9",
      accentSoft: "rgba(56, 189, 248, 0.12)",
      accentForeground: "#04121b",
      logoGlyph: "A",
    },
    features: { leaderboard: true, journal: true, payouts: false },
    program: {
      currency: "USD",
      accountSizes: [25_000, 50_000, 200_000],
      profitTargetPct: 8,
      maxDailyLossPct: 4,
      maxDrawdownPct: 8,
      drawdownMode: "trailing",
      profitSplitPct: 90,
      maxStakePerOrder: 5_000,
      maxExposurePerMarket: 10_000,
      challengeDurationDays: 45,
      minTradingDays: 7,
    },
  },
  {
    slug: "nova",
    clientKey: "nova",
    name: "Nova Markets",
    tagline: "Predict boldly. Trade funded.",
    branding: {
      accent: "#a78bfa",
      accentHover: "#8b5cf6",
      accentSoft: "rgba(167, 139, 250, 0.12)",
      accentForeground: "#120a24",
      logoGlyph: "N",
    },
    features: { leaderboard: true, journal: false, payouts: true },
    program: {
      currency: "USD",
      accountSizes: [10_000, 50_000, 100_000],
      profitTargetPct: 12,
      maxDailyLossPct: 5,
      maxDrawdownPct: 12,
      drawdownMode: "absolute",
      profitSplitPct: 85,
      maxStakePerOrder: 3_000,
      maxExposurePerMarket: 6_000,
      challengeDurationDays: 30,
      minTradingDays: 5,
    },
  },
] as const;

let seeded = false;

async function ensureWebhookApiKeys(): Promise<void> {
  const tenants = await prisma.tenant.findMany({ select: { id: true, slug: true } });
  for (const tenant of tenants) {
    const count = await prisma.propFirmApiKey.count({ where: { tenantId: tenant.id } });
    if (count > 0) continue;

    const key = await createPropFirmApiKey(tenant.id, "webhook");
    if (process.env.NODE_ENV !== "production") {
      console.info(`[seed] Webhook API key for ${tenant.slug}: ${key.rawKey}`);
    }
  }
}

export async function ensureSeeded(): Promise<void> {
  if (seeded) {
    if (process.env.DATABASE_URL) {
      await ensureWebhookApiKeys();
      await ensureFirmSettingsSeeded();
    }
    return;
  }
  if (!process.env.DATABASE_URL) return;

  try {
    const count = await prisma.tenant.count();
    if (count > 0) {
      seeded = true;
      await ensureWebhookApiKeys();
      await ensureFirmSettingsSeeded();
      return;
    }

    const passwordHash = await hashPassword(DEMO_PASSWORD);

    for (const tenant of SEED_TENANTS) {
      const row = await prisma.tenant.create({
        data: {
          slug: tenant.slug,
          clientKey: tenant.clientKey,
          name: tenant.name,
          tagline: tenant.tagline,
          branding: tenant.branding,
          features: tenant.features,
          program: tenant.program,
        },
      });

      await prisma.user.createMany({
        data: [
          {
            tenantId: row.id,
            email: `trader@${tenant.slug}.demo`,
            displayName: "Demo Trader",
            hashedPassword: passwordHash,
            role: "trader",
          },
          {
            tenantId: row.id,
            email: `admin@${tenant.slug}.demo`,
            displayName: "Firm Admin",
            hashedPassword: passwordHash,
            role: "prop_firm_admin",
          },
        ],
      });
    }

    await prisma.user.create({
      data: {
        tenantId: null,
        email: "super@proppredict.demo",
        displayName: "Super Admin",
        hashedPassword: passwordHash,
        role: "super_admin",
      },
    });

    seeded = true;
    await ensureWebhookApiKeys();
    await ensureFirmSettingsSeeded();
  } catch (error) {
    console.error("[seed] failed:", error);
  }
}
